source/grid/tria.cc

```diff
@@ -15165,6 +15168,157 @@ Triangulation<1, 3>::prepare_coarsening_and_refinement()
+  // vertices we encounter.
+  //
+  // This scheme ensures that for loops over active cells we encounter
+  // vertices in their natural order, whereas for other loops we see
+  // them in one of two blocks, each of which we traverse in consecutive
+  // order.
+  std::vector<unsigned int> old_to_new_vertex_indices(
+    vertices.size(), numbers::invalid_unsigned_int);
+  {
+    // We will access the indices of vertices of cells below. For
+    // this to work, we have to first make sure the cache for this
+    // information is up to date.
+    reset_cell_vertex_indices_cache();
+
+    unsigned int next_unused_index = 0;
+
+    // First the active cell iterators:
+    for (const auto &cell : active_cell_iterators())
+      for (const unsigned int v : cell->vertex_indices())
+        {
+          const unsigned int vertex_index = cell->vertex_index(v);
+          if (old_to_new_vertex_indices[vertex_index] ==
+              numbers::invalid_unsigned_int)
+            old_to_new_vertex_indices[vertex_index] = next_unused_index++;
+        }
+    // Then the non-active cell iterators (we already covered the active
+    // ones):
+    for (const auto &cell : cell_iterators())
+      if (cell->is_active() == false)
+        for (const unsigned int v : cell->vertex_indices())
+          {
+            const unsigned int vertex_index = cell->vertex_index(v);
+            if (old_to_new_vertex_indices[vertex_index] ==
+                numbers::invalid_unsigned_int)
+              old_to_new_vertex_indices[vertex_index] = next_unused_index++;
+          }
+
+    // At this point, we should have covered all used vertices. Check this:
+    for (unsigned int v = 0; v < vertices.size(); ++v)
+      Assert(((vertices_used[v] == true) && (old_to_new_vertex_indices[v] !=
+                                             numbers::invalid_unsigned_int)) ||
+               ((vertices_used[v] == false) && (old_to_new_vertex_indices[v] ==
+                                                numbers::invalid_unsigned_int)),
+             ExcInternalError());
+
+    // This leaves the unused vertices. Enumerate them as well:
+    for (unsigned int v = 0; v < vertices.size(); ++v)
+      if (old_to_new_vertex_indices[v] == numbers::invalid_unsigned_int)
+        old_to_new_vertex_indices[v] = next_unused_index++;
+    Assert(next_unused_index == vertices.size(), ExcInternalError());
+  }
+
+
+  // Step 2: Now we need to reorder or reindex all sorts of data structures.
+  // This can all be done on parallel tasks.
+  Threads::TaskGroup<> tasks;
+
+  // Reorder the vertices[] array:
+  tasks += Threads::new_task(
+    [&old_to_new_vertex_indices, &vertices = this->vertices]() {
+      std::vector<Point<spacedim>> new_vertices(vertices.size());
+      for (unsigned int v = 0; v < old_to_new_vertex_indices.size(); ++v)
+        new_vertices[old_to_new_vertex_indices[v]] = vertices[v];
+      vertices = std::move(new_vertices);
+    });
```

Comment:
*mhm* I am not sure that running all of these reordering on tasks has any net benefit at all.

All of the tasks do a simple copy of data with a lookup of a new index. Having them run on multiple tasks that potentially get rescheduled over multiple different cores just sounds like the whole operation taking more time than less.

Would it be OK to simply do all of the reordering sequentally for the time being?

Reply:
That's perhaps a discussion to have once we agree that this patch is the right way to go. I want to use tasks much more widely in deal.II -- we all have many cores on our laptops, and we're not using them sufficiently well.

In the meantime, these arrays all have tens or hundreds of thousands of entries. They're also all separate, so need to be loaded into the cache of one core or another anyway. Surely doing this for arrays of this size is worth doing on multiple cores at the same time.